File: hackerrank/longest_increasing_subsequence.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstring>
using namespace std;

void find_cur_longest(vector<int> arr, vector<int> *LIS, vector<vector<int>> *sub_LIS) {
}

// Find largest value smaller than val
int binary_search(int arr[], int val, int left, int right) {
    // while (right - left > 1) {
    //     int mid_i = (left+right)/2;
    //     if (arr[mid_i] >= val) {
    //         right = mid_i;
    //     } else {
    //         left = mid_i;
    //     }
    // }

    // return right;

    if (right - left <= 1) {
        return right;
    }

    int mid = (right+left)/2;
    if (arr[mid] >= val) {
        right = mid;
        return binary_search(arr, val, left, right);
    } else {
        left = mid;
        return binary_search(arr, val, left, right);
    }
}

void LIS(int arr[], int arr_size) {

    // Value at index 'i' is larger value in sub LIS of length 'i'
    int sub_LIS[arr_size];
    for (int i = 0; i < arr_size; ++i) {
        sub_LIS[i] = 0;
    }

    int LIS = 1;
    sub_LIS[0] = arr[0];

    // Go through each number in array
    for (int i = 1; i < arr_size; ++i) {
        int num = arr[i];

        // If number is smaller than smallest,
        //  replace it
        if (num < sub_LIS[0]) {
            sub_LIS[0] = num;

        // If number is larger than largest, extend list
        } else if (num > sub_LIS[LIS-1]) {
            sub_LIS[LIS++] = num;

        } else {

        // Otherwise, find largest number smaller than current num,
        //  and replace the sub LIS that is one longer than that
            int index = binary_search(sub_LIS, num, 0, LIS-1);
            sub_LIS[index] = num;
        }
        // for (int i : sub_LIS) {
        //     cout << i << " ";
        // } cout << endl;
    }

    // for (int i : sub_LIS) {
    //     cout << i << " ";
    // } cout << endl;
    cout << LIS << endl;
}

int main() {
    int n;
    cin >> n;
    int arr[n];
    for (int i = 0; i < n; i++) {
        int num;
        cin >> num;
        arr[i] = num;
    }

    LIS(arr, n);
}

File: hackerrank/longest_increasing_subsequence.py
def longest_increasing_subsequence(arr):

    sub_LIS = []
    longest_LIS = 0;

    # Go through each number in array
    for num in arr:

        # Initial longest list is simply the current number
        cur_longest_sub_LIS = 0

        # Find the longest previous LIS whose last value is < num
        for i, sub_LIS_arr in enumerate(sub_LIS):

            # For a particular previous LIS, if its last value is smaller than num and length greater than current LIS
            # if len(sub_LIS_arr) > 0 and sub_LIS_arr[-1] <= num and len(sub_LIS_arr) > len(cur_longest_sub_LIS):
            if arr[i] <= num and sub_LIS_arr > cur_longest_sub_LIS:
                # Then update the new LIS
                cur_longest_sub_LIS = sub_LIS_arr

        # Finish processing LIS ending with current num
        # if (len(cur_longest_sub_LIS) == 0):
        if cur_longest_sub_LIS == 0:
            #cur_longest_sub_LIS = [num]
            cur_longest_sub_LIS = 1
        else:
            # cur_longest_sub_LIS.append(num)
            cur_longest_sub_LIS += 1

        sub_LIS.append(cur_longest_sub_LIS)
        # if len(cur_longest_sub_LIS) > len(longest_LIS):
        if cur_longest_sub_LIS > longest_LIS:
            longest_LIS = cur_longest_sub_LIS
    return longest_LIS

c = int(input())
arr = []
for x in range(0, c):
    arr.append(int(input()))

print ( longest_increasing_subsequence(arr)  )
